File: src/claude/runner.js
import { spawn, execSync } from 'child_process';
import { platform } from 'os';
import { existsSync, mkdirSync, copyFileSync, rmSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { buildPrompt } from './prompt.js';
import { handleClaudeEvent } from './events.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const UPLOADS_DIR = join(__dirname, '..', '..', 'uploads');

const activeProcesses = new Map();
const activeToolCalls = new Map();
const taskUsage = new Map();
const startingTasks = new Set();

const IS_WIN = platform() === 'win32';
const MODEL_MAP = { opus: 'opus', sonnet: 'sonnet', haiku: 'haiku' };

// Server-side logger
const CLR = {
  reset: '\x1b[0m',
  gray: '\x1b[90m',
  magenta: '\x1b[35m',
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  cyan: '\x1b[36m',
};
function ts() {
  return `${CLR.gray}${new Date().toLocaleTimeString('tr-TR', { hour12: false })}${CLR.reset}`;
}
const slog = {
  claude: (msg) => console.log(`${ts()} ${CLR.magenta}[CLAUDE]${CLR.reset} ${msg}`),
  task: (msg) => console.log(`${ts()} ${CLR.yellow}[TASK]${CLR.reset} ${msg}`),
  git: (msg) => console.log(`${ts()} ${CLR.blue}[GIT]${CLR.reset} ${msg}`),
};

// Generate a short, meaningful branch name from task title
function generateBranchSlug(title) {
  return title
    .toLowerCase()
    .replace(/[çÇ]/g, 'c')
    .replace(/[ğĞ]/g, 'g')
    .replace(/[ıİ]/g, 'i')
    .replace(/[öÖ]/g, 'o')
    .replace(/[şŞ]/g, 's')
    .replace(/[üÜ]/g, 'u')
    .replace(/[^a-z0-9\s-]/g, '')
    .trim()
    .replace(/\s+/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '')
    .slice(0, 40)
    .replace(/-$/, '');
}

// Auto-create git branch for a task before Claude starts
function ensureTaskBranch(task, workingDir, project, queries, io) {
  if (!project.auto_branch) return null;
  const isRevision = (task.revision_count || 0) > 0;
  const slug = generateBranchSlug(task.title) || `task-${task.id}`;
  const branchName = task.branch_name || `${task.task_type || 'feature'}/${slug}`;

  try {
    const exec = (cmd) => execSync(cmd, { cwd: workingDir, stdio: 'pipe', timeout: 10000 }).toString().trim();

    // Check if we're in a git repo
    exec('git rev-parse --is-inside-work-tree');

    if (isRevision && task.branch_name) {
      // For revisions, just checkout the existing branch
      try {
        const currentBranch = exec('git branch --show-current');
        if (currentBranch !== task.branch_name) {
          exec(`git checkout ${task.branch_name}`);
          addLog(task.id, `Switched to branch: ${task.branch_name}`, 'system', queries, io);
        }
      } catch {
        addLog(task.id, `Branch ${task.branch_name} not found, creating fresh`, 'info', queries, io);
        exec(`git checkout -b ${branchName}`);
      }
    } else {
      // For new tasks, create and checkout branch
      try {
        // Determine base branch
        const baseBranch = project.pr_base_branch || 'main';
        try {
          exec(`git rev-parse --verify ${baseBranch}`);
        } catch {
          // fallback: stay on current branch as base
        }

        // Check if branch already exists
        try {
          exec(`git rev-parse --verify ${branchName}`);
          // Branch exists, just checkout
          exec(`git checkout ${branchName}`);
          addLog(task.id, `Switched to existing branch: ${branchName}`, 'system', queries, io);
        } catch {
          // Branch doesn't exist, create from base
          try {
            exec(`git checkout -b ${branchName} ${baseBranch}`);
          } catch {
            exec(`git checkout -b ${branchName}`);
          }
          addLog(task.id, `Created branch: ${branchName}`, 'system', queries, io);
        }
      } catch (e) {
        addLog(task.id, `Git branch setup failed: ${e.message}`, 'error', queries, io);
        return null;
      }
    }

    // Store branch name in task
    queries.updateTaskBranch.run(branchName, task.id);
    return branchName;
  } catch {
    return null;
  }
}

// Auto-create PR after task completes
function createPullRequest(task, workingDir, project, queries, io) {
  if (!project.auto_pr || !task.branch_name) return null;

  try {
    const exec = (cmd) => execSync(cmd, { cwd: workingDir, stdio: 'pipe', timeout: 15000 }).toString().trim();
    const baseBranch = project.pr_base_branch || 'main';
    const branch = task.branch_name;

    // Check if PR already exists
    try {
      const existing = exec(`gh pr view ${branch} --json url --jq .url`);
      if (existing.startsWith('http')) {
        addLog(task.id, `PR already exists: ${existing}`, 'system', queries, io);
        return existing;
      }
    } catch {
      // No existing PR, create one
    }

    // Check if there are commits to PR
    try {
      const diff = exec(`git log ${baseBranch}..${branch} --oneline`);
      if (!diff) {
        addLog(task.id, 'No commits to create PR for', 'info', queries, io);
        return null;
      }
    } catch {
      return null;
    }

    // Create PR
    const title = task.title;
    const body = `## Task #${task.id}: ${task.title}\n\n${task.description || 'No description provided.'}\n\n---\n*Auto-generated by Claude Board*`;
    const prUrl = exec(
      `gh pr create --base ${baseBranch} --head ${branch} --title "${title.replace(/"/g, '\\"')}" --body "${body.replace(/"/g, '\\"')}"`,
    );

    if (prUrl.startsWith('http')) {
      addLog(task.id, `PR created: ${prUrl}`, 'success', queries, io);
      queries.updateTaskGitInfo.run(task.commits || '[]', prUrl, task.diff_stat, task.id);
      return prUrl;
    }
    return null;
  } catch (e) {
    addLog(task.id, `Auto-PR failed: ${e.message}`, 'error', queries, io);
    return null;
  }
}

// Cross-platform: resolve claude executable
function getClaudeCommand() {
  // On Windows, spawn with shell:false needs exact executable.
  // 'claude' works if it's a .exe in PATH.
  // Some installs use .cmd wrapper — shell:true handles that but has escaping issues.
  // We use shell:false with the bare command; Node resolves .exe on Windows via PATHEXT.
  return 'claude';
}

// Cross-platform process kill
function killProcess(proc) {
  try {
    if (IS_WIN) {
      // Windows: SIGTERM doesn't reliably kill child trees. Use taskkill.
      spawn('taskkill', ['/pid', String(proc.pid), '/T', '/F'], { stdio: 'ignore', windowsHide: true });
    } else {
      // Unix: send SIGTERM to process group
      proc.kill('SIGTERM');
    }
  } catch {
    try {
      proc.kill('SIGKILL');
    } catch {}
  }
}

export function isTaskRunning(taskId) {
  return activeProcesses.has(taskId);
}
export function getActiveProcess(taskId) {
  return activeProcesses.get(taskId);
}

export function stopClaude(taskId, io, queries) {
  const proc = activeProcesses.get(taskId);
  if (proc) {
    slog.claude(`Stopping task #${taskId} (PID: ${proc.pid})`);
    killProcess(proc);
    activeProcesses.delete(taskId);
    startingTasks.delete(taskId);
    taskUsage.delete(taskId);
    addLog(taskId, 'Claude process stopped by user.', 'system', queries, io);
  }
}

function addLog(taskId, message, logType, queries, io, meta = null) {
  try {
    queries.addTaskLog.run(taskId, message, logType, meta);
  } catch (e) {
    console.error('[Log]', e.message);
  }
  const payload = { taskId, message, logType, created_at: new Date().toISOString() };
  if (meta) payload.meta = meta;
  io.emit('task:log', payload);
}

// Scan git for recent commits and PR URLs after task completes
function scanGitInfo(workingDir, taskId, queries) {
  try {
    const exec = (cmd) => execSync(cmd, { cwd: workingDir, stdio: 'pipe', timeout: 5000 }).toString().trim();

    // Get recent commits (last 10, on current branch)
    const logOutput = exec('git log --oneline -10 --no-merges --format="%H|%h|%s|%an|%ai"');
    const commits = logOutput
      .split('\n')
      .filter(Boolean)
      .map((line) => {
        const [hash, short, message, author, date] = line.split('|');
        return { hash, short, message, author, date };
      });

    // Try to find remote URL for commit links
    let repoUrl = '';
    try {
      const remote = exec('git remote get-url origin');
      repoUrl = remote
        .replace(/\.git$/, '')
        .replace(/^git@github\.com:/, 'https://github.com/')
        .replace(/^git@(.+):/, 'https://$1/');
    } catch {}

    if (repoUrl) {
      commits.forEach((c) => {
        c.url = `${repoUrl}/commit/${c.hash}`;
      });
    }

    // Capture diff stat
    let diffStat = null;
    try {
      const branch = exec('git branch --show-current');
      if (branch && branch !== 'main' && branch !== 'master') {
        // Diff against the base branch (main or master)
        let baseBranch = 'main';
        try {
          exec('git rev-parse --verify main');
        } catch {
          baseBranch = 'master';
        }
        diffStat = exec(`git diff --stat ${baseBranch}...HEAD`);
      } else if (commits.length > 0) {
        diffStat = exec(`git diff --stat HEAD~${Math.min(commits.length, 10)}..HEAD`);
      }
    } catch {}

    // Try to find PR URL (if gh CLI is available)
    let prUrl = null;
    try {
      const branch = exec('git branch --show-current');
      if (branch && branch !== 'main' && branch !== 'master') {
        const prOutput = exec(`gh pr view ${branch} --json url --jq .url`);
        if (prOutput.startsWith('http')) prUrl = prOutput;
      }
    } catch {}

    queries.updateTaskGitInfo.run(commits, prUrl, diffStat, taskId);
    return { commits, prUrl, diffStat };
  } catch {
    return { commits: [], prUrl: null };
  }
}

export function startClaude(
  task,
  io,
  workingDir,
  project = {},
  revisions = [],
  snippets = [],
  { queries, statsQueries, activityLog, onFinished, attachments = [], role = null } = {},
) {
  if (activeProcesses.has(task.id) || startingTasks.has(task.id)) {
    addLog(task.id, 'Claude is already running for this task.', 'system', queries, io);
    return;
  }
  startingTasks.add(task.id);
  slog.task(`Starting #${task.id} "${task.title}" [${task.task_type}]`);

  // Copy attachments to working directory
  const attachDir = join(workingDir, '.claude-attachments');
  if (attachments.length > 0) {
    try {
      if (!existsSync(attachDir)) mkdirSync(attachDir, { recursive: true });
      for (const a of attachments) {
        const src = join(UPLOADS_DIR, a.filename);
        const dest = join(attachDir, a.filename);
        if (existsSync(src)) {
          copyFileSync(src, dest);
          slog.task(`Attached: ${a.original_name} → .claude-attachments/${a.filename}`);
        }
      }
    } catch (e) {
      slog.task(`${CLR.red}Attachment copy failed:${CLR.reset} ${e.message}`);
    }
  }

  const prompt = buildPrompt(task, revisions, snippets, attachments, role);
  const model = task.model || 'sonnet';
  const effort = task.thinking_effort || 'medium';
  const permissionMode = project.permission_mode || 'auto-accept';
  slog.claude(`Config: model=${model} effort=${effort} permissions=${permissionMode}`);

  // Snapshot baseline usage for live tracking
  const currentTask = queries.getTaskById.get(task.id);
  taskUsage.set(task.id, {
    baseline: {
      input: currentTask?.input_tokens || 0,
      output: currentTask?.output_tokens || 0,
      cacheRead: currentTask?.cache_read_tokens || 0,
      cacheCreation: currentTask?.cache_creation_tokens || 0,
      cost: currentTask?.total_cost || 0,
    },
    session: { input: 0, output: 0, cacheRead: 0, cacheCreation: 0 },
  });

  // Auto-create git branch if enabled
  const branchName = ensureTaskBranch(task, workingDir, project, queries, io);
  if (branchName) {
    task.branch_name = branchName;
    slog.git(`Task #${task.id} → branch: ${branchName}`);
  }

  addLog(task.id, `Starting Claude for task: ${task.title}`, 'system', queries, io);
  addLog(
    task.id,
    `Model: ${model} | Effort: ${effort} | Permissions: ${permissionMode}${task.branch_name ? ` | Branch: ${task.branch_name}` : ''}`,
    'info',
    queries,
    io,
  );

  if (activityLog) {
    activityLog.add(task.project_id, task.id, 'claude_started', `Claude started: ${task.title}`, { model, effort });
  }

  const args = ['-p', prompt, '--output-format', 'stream-json', '--verbose'];
  if (MODEL_MAP[model]) args.push('--model', MODEL_MAP[model]);

  if (permissionMode === 'auto-accept') {
    args.push('--dangerously-skip-permissions');
  } else if (permissionMode === 'allow-tools') {
    const tools = (project.allowed_tools || '')
      .split(',')
      .map((t) => t.trim())
      .filter(Boolean);
    if (tools.length > 0) {
      for (const t of tools) args.push('--allowedTools', t);
    } else {
      args.push('--dangerously-skip-permissions');
    }
  }

  const proc = spawn(getClaudeCommand(), args, {
    cwd: workingDir,
    shell: false,
    env: { ...process.env },
    stdio: ['ignore', 'pipe', 'pipe'],
    ...(IS_WIN && { windowsHide: true }),
  });

  activeProcesses.set(task.id, proc);
  startingTasks.delete(task.id);
  slog.claude(`Task #${task.id} spawned (PID: ${proc.pid}) — ${activeProcesses.size} active process(es)`);
  let buffer = '';

  const taskAddLog = (tid, msg, type, meta = null) => addLog(tid, msg, type, queries, io, meta);

  proc.stdout.on('data', (data) => {
    buffer += data.toString();
    const lines = buffer.split('\n');
    buffer = lines.pop();
    for (const line of lines) {
      if (!line.trim()) continue;
      try {
        const event = JSON.parse(line);
        handleClaudeEvent(task.id, event, {
          addLog: taskAddLog,
          queries,
          statsQueries,
          io,
          taskUsage,
          activeToolCalls,
        });
      } catch {
        taskAddLog(task.id, line, 'claude');
      }
    }
  });

  proc.stderr.on('data', (data) => {
    const msg = data.toString().trim();
    if (!msg) return;
    if (
      msg.toLowerCase().includes('rate limit') ||
      msg.toLowerCase().includes('429') ||
      msg.toLowerCase().includes('overloaded')
    ) {
      queries.incrementRateLimitHits.run(task.id);
      taskAddLog(task.id, `Rate limit hit: ${msg}`, 'error');
    } else {
      taskAddLog(task.id, msg, 'error');
    }
  });

  proc.on('close', (code) => {
    activeProcesses.delete(task.id);
    startingTasks.delete(task.id);
    taskUsage.delete(task.id);

    if (code === 0) {
      // Scan git for commits and PRs
      const gitInfo = scanGitInfo(workingDir, task.id, queries);
      if (gitInfo.commits.length > 0) {
        const commitCount = gitInfo.commits.length;
        const prInfo = gitInfo.prUrl ? ` | PR: ${gitInfo.prUrl}` : '';
        taskAddLog(task.id, `Git: ${commitCount} commit(s) found${prInfo}`, 'system');
        slog.git(`Task #${task.id}: ${commitCount} commit(s)${prInfo}`);
      }

      // Auto-create PR if enabled
      const updatedTask = queries.getTaskById.get(task.id);
      if (project.auto_pr && updatedTask.branch_name) {
        slog.git(`Task #${task.id}: Creating PR for ${updatedTask.branch_name}`);
        createPullRequest(updatedTask, workingDir, project, queries, io);
      }

      // Calculate duration
      const taskData = queries.getTaskById.get(task.id);
      let durationStr = '';
      if (taskData?.started_at) {
        const dur = Math.round((Date.now() - new Date(taskData.started_at).getTime()) / 1000);
        const m = Math.floor(dur / 60);
        const s = dur % 60;
        durationStr = m > 0 ? ` in ${m}m ${s}s` : ` in ${s}s`;
      }
      const tokens = (taskData?.input_tokens || 0) + (taskData?.output_tokens || 0);
      const cost = taskData?.total_cost || 0;
      slog.task(
        `${CLR.green}✓${CLR.reset} #${task.id} "${task.title}" completed${durationStr} | ${tokens.toLocaleString()} tokens | $${cost.toFixed(4)}`,
      );

      taskAddLog(task.id, 'Claude finished successfully.', 'success');
      queries.updateTaskStatus.run('testing', task.id);
      queries.setTaskCompleted.run(task.id);
      const updated = queries.getTaskById.get(task.id);
      io.emit('task:updated', updated);
      if (activityLog) activityLog.add(task.project_id, task.id, 'task_completed', `Task completed: ${task.title}`);
    } else {
      slog.task(`${CLR.red}✗${CLR.reset} #${task.id} "${task.title}" failed (exit code: ${code})`);
      taskAddLog(task.id, `Claude exited with code ${code}.`, 'error');
      if (activityLog)
        activityLog.add(task.project_id, task.id, 'task_failed', `Task failed (exit ${code}): ${task.title}`);
    }

    // Cleanup attachments dir
    if (attachments.length > 0 && existsSync(attachDir)) {
      try {
        rmSync(attachDir, { recursive: true, force: true });
      } catch {}
    }

    io.emit('claude:finished', { taskId: task.id, exitCode: code });
    if (onFinished) onFinished(task, code);
  });

  proc.on('error', (err) => {
    activeProcesses.delete(task.id);
    startingTasks.delete(task.id);
    taskUsage.delete(task.id);
    slog.claude(`${CLR.red}ERROR${CLR.reset} Task #${task.id}: Failed to start — ${err.message}`);
    taskAddLog(task.id, `Failed to start Claude: ${err.message}`, 'error');
    io.emit('claude:finished', { taskId: task.id, exitCode: -1 });
    if (activityLog) activityLog.add(task.project_id, task.id, 'task_failed', `Failed to start: ${err.message}`);
  });
}

File: src-tauri/src/claude/runner.rs
use std::collections::{HashMap, HashSet};
use std::io::{BufRead, BufReader};
use std::path::Path;
use std::process::{Command, Stdio};
use std::sync::{Arc, Mutex};
use tauri::{AppHandle, Emitter};
use crate::db::{self, DbPool};
use crate::db::{tasks, snippets, attachments, roles, projects, activity};
use super::events::{EventContext, UsageTracker, UsageBaseline, UsageSession};
use super::prompt::build_prompt;

#[cfg(target_os = "windows")]
use std::os::windows::process::CommandExt;
#[cfg(target_os = "windows")]
const CREATE_NO_WINDOW: u32 = 0x08000000;

type ProcessMap = Arc<Mutex<HashMap<i64, u32>>>;
type StartingSet = Arc<Mutex<HashSet<i64>>>;

static ACTIVE_PROCESSES: once_cell::sync::Lazy<ProcessMap> = once_cell::sync::Lazy::new(|| Arc::new(Mutex::new(HashMap::new())));
static STARTING_TASKS: once_cell::sync::Lazy<StartingSet> = once_cell::sync::Lazy::new(|| Arc::new(Mutex::new(HashSet::new())));
static EVENT_CTX: once_cell::sync::Lazy<EventContext> = once_cell::sync::Lazy::new(EventContext::new);

pub fn is_running(task_id: i64) -> bool {
    ACTIVE_PROCESSES.lock().unwrap().contains_key(&task_id)
}

pub fn stop(task_id: i64, db: &DbPool, app: &AppHandle) {
    if let Some(pid) = ACTIVE_PROCESSES.lock().unwrap().remove(&task_id) {
        kill_process(pid);
        STARTING_TASKS.lock().unwrap().remove(&task_id);
        EVENT_CTX.task_usage.lock().unwrap().remove(&task_id);
        tasks::add_log(db, task_id, "Claude process stopped by user.", "system", None);
        app.emit("task:log", &serde_json::json!({
            "taskId": task_id, "message": "Claude process stopped by user.", "logType": "system"
        })).ok();
    }
}

fn kill_process(pid: u32) {
    #[cfg(target_os = "windows")]
    {
        Command::new("taskkill")
            .args(["/pid", &pid.to_string(), "/T", "/F"])
            .stdout(Stdio::null()).stderr(Stdio::null())
            .creation_flags(CREATE_NO_WINDOW)
            .spawn().ok();
    }
    #[cfg(not(target_os = "windows"))]
    {
        unsafe { libc::kill(pid as i32, libc::SIGTERM); }
    }
}

fn generate_branch_slug(title: &str) -> String {
    title
        .to_lowercase()
        .replace(['ç', 'Ç'], "c").replace(['ğ', 'Ğ'], "g")
        .replace(['ı', 'İ'], "i").replace(['ö', 'Ö'], "o")
        .replace(['ş', 'Ş'], "s").replace(['ü', 'Ü'], "u")
        .chars().filter(|c| c.is_alphanumeric() || c.is_whitespace() || *c == '-').collect::<String>()
        .trim().replace(char::is_whitespace, "-")
        .replace("--", "-").trim_matches('-').to_string()
        .chars().take(40).collect::<String>()
        .trim_end_matches('-').to_string()
}

fn ensure_task_branch(task: &tasks::Task, working_dir: &str, project: &projects::Project, db: &DbPool, app: &AppHandle) -> Option<String> {
    if project.auto_branch.unwrap_or(1) == 0 { return None; }
    let is_revision = task.revision_count.unwrap_or(0) > 0;
    let slug = generate_branch_slug(&task.title);
    let slug = if slug.is_empty() { format!("task-{}", task.id) } else { slug };
    let branch_name = task.branch_name.clone().unwrap_or_else(|| {
        format!("{}/{}", task.task_type.as_deref().unwrap_or("feature"), slug)
    });

    let exec = |cmd: &str| -> Result<String, String> {
        let mut c = Command::new("git");
        c.args(cmd.split_whitespace())
            .current_dir(working_dir)
            .stdout(Stdio::piped()).stderr(Stdio::piped());
        #[cfg(target_os = "windows")]
        c.creation_flags(CREATE_NO_WINDOW);
        let output = c.output().map_err(|e| e.to_string())?;
        if output.status.success() {
            Ok(String::from_utf8_lossy(&output.stdout).trim().to_string())
        } else {
            Err(String::from_utf8_lossy(&output.stderr).trim().to_string())
        }
    };

    if exec("rev-parse --is-inside-work-tree").is_err() { return None; }

    let git_hidden = |args: &[&str], dir: &str| -> std::io::Result<std::process::Output> {
        let mut c = Command::new("git");
        c.args(args).current_dir(dir).stdout(Stdio::piped()).stderr(Stdio::piped());
        #[cfg(target_os = "windows")]
        c.creation_flags(CREATE_NO_WINDOW);
        c.output()
    };

    if is_revision && task.branch_name.is_some() {
        let bn = task.branch_name.as_deref().unwrap();
        if let Ok(current) = exec("branch --show-current") {
            if current != bn {
                let _ = git_hidden(&["checkout", bn], working_dir);
            }
        }
    } else {
        let base = project.pr_base_branch.as_deref().unwrap_or("main");
        if exec(&format!("rev-parse --verify {}", branch_name)).is_ok() {
            let _ = git_hidden(&["checkout", &branch_name], working_dir);
        } else if git_hidden(&["checkout", "-b", &branch_name, base], working_dir).is_err() {
            let _ = git_hidden(&["checkout", "-b", &branch_name], working_dir);
        }
    }

    tasks::update_branch(db, task.id, &branch_name);
    let _ = app;
    Some(branch_name)
}

fn scan_git_info(working_dir: &str, task_id: i64, db: &DbPool) {
    let exec = |args: &[&str]| -> Option<String> {
        let mut cmd = Command::new("git");
        cmd.args(args).current_dir(working_dir)
            .stdout(Stdio::piped()).stderr(Stdio::null());
        #[cfg(target_os = "windows")]
        cmd.creation_flags(CREATE_NO_WINDOW);
        cmd.output().ok()
            .filter(|o| o.status.success())
            .map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string())
    };

    let log_output = exec(&["log", "--oneline", "-10", "--no-merges", "--format=%H|%h|%s|%an|%ai"]).unwrap_or_default();
    let commits: Vec<serde_json::Value> = log_output.lines().filter(|l| !l.is_empty()).map(|line| {
        let parts: Vec<&str> = line.splitn(5, '|').collect();
        serde_json::json!({
            "hash": parts.first().unwrap_or(&""),
            "short": parts.get(1).unwrap_or(&""),
            "message": parts.get(2).unwrap_or(&""),
            "author": parts.get(3).unwrap_or(&""),
            "date": parts.get(4).unwrap_or(&""),
        })
    }).collect();

    let diff_stat = exec(&["diff", "--stat", "HEAD~1..HEAD"]);
    let pr_url = exec(&["branch", "--show-current"]).and_then(|branch| {
        if branch == "main" || branch == "master" { return None; }
        exec(&["gh", "pr", "view", &branch, "--json", "url", "--jq", ".url"])
            .filter(|u| u.starts_with("http"))
    });

    let commits_json = serde_json::to_string(&commits).unwrap_or_else(|_| "[]".into());
    tasks::update_git_info(db, task_id, &commits_json, pr_url.as_deref(), diff_stat.as_deref());
}

/// Copy task attachments from uploads dir to working dir for Claude access.
fn copy_task_attachments(task_id: i64, working_dir: &str, db: &DbPool) -> (Vec<attachments::Attachment>, std::path::PathBuf) {
    let task_attachments = attachments::get_by_task(db, task_id);
    let uploads_dir = db::get_data_dir().parent().map(|p| p.join("uploads")).unwrap_or_default();
    let attach_dir = Path::new(working_dir).join(".claude-attachments");

    if !task_attachments.is_empty() {
        std::fs::create_dir_all(&attach_dir).ok();
        for a in &task_attachments {
            let src = uploads_dir.join(&a.filename);
            let dest = attach_dir.join(&a.filename);
            if src.exists() { std::fs::copy(&src, &dest).ok(); }
        }
    }

    (task_attachments, attach_dir)
}

/// Build Claude CLI arguments from task configuration.
fn build_claude_args(
    prompt: &str,
    model: &str,
    effort: &str,
    permission_mode: &str,
    allowed_tools: &str,
    mcp_server_port: u16,
) -> Vec<String> {
    let mut args = vec![
        "-p".to_string(), prompt.to_string(),
        "--output-format".to_string(), "stream-json".to_string(),
        "--verbose".to_string(),
        "--model".to_string(), model.to_string(),
    ];

    // MCP config
    let mcp_server_path = std::env::current_exe()
        .ok()
        .and_then(|p| p.parent().map(|d| d.join("mcp-server.js").to_string_lossy().to_string()))
        .unwrap_or_default();

    let mcp_config = serde_json::json!({
        "mcpServers": {
            "claude-board": {
                "command": "node",
                "args": [mcp_server_path],
                "env": { "CLAUDE_BOARD_URL": format!("http://localhost:{}", mcp_server_port) }
            }
        }
    });
    args.extend(["--mcp-config".to_string(), mcp_config.to_string()]);

    // Permission mode
    if permission_mode == "auto-accept" {
        args.push("--dangerously-skip-permissions".to_string());
    } else if permission_mode == "allow-tools" {
        let tools: Vec<&str> = allowed_tools.split(',').map(|t| t.trim()).filter(|t| !t.is_empty()).collect();
        if tools.is_empty() {
            args.push("--dangerously-skip-permissions".to_string());
        } else {
            for t in tools { args.extend(["--allowedTools".to_string(), t.to_string()]); }
        }
    }

    if effort != "medium" {
        args.extend(["--thinking-budget".to_string(), effort.to_string()]);
    }

    args
}

/// Handle process output, track events, and update task state on completion.
fn handle_process_lifecycle(
    task_id: i64,
    mut child: std::process::Child,
    db: &DbPool,
    app: &AppHandle,
    working_dir: &str,
    project_id: i64,
    task_title: &str,
    attach_dir: &Path,
) {
    let pid = child.id();
    ACTIVE_PROCESSES.lock().unwrap().insert(task_id, pid);
    STARTING_TASKS.lock().unwrap().remove(&task_id);

    // Read stdout (safe: we configured Stdio::piped)
    if let Some(stdout) = child.stdout.take() {
        let reader = BufReader::new(stdout);
        for line in reader.lines().flatten() {
            if line.trim().is_empty() { continue; }
            match serde_json::from_str::<serde_json::Value>(&line) {
                Ok(event) => super::events::handle_event(task_id, &event, db, app, &EVENT_CTX),
                Err(_) => { tasks::add_log(db, task_id, &line, "claude", None); }
            }
        }
    }

    let status = child.wait().ok().and_then(|s| s.code()).unwrap_or(-1);

    // Cleanup process tracking
    ACTIVE_PROCESSES.lock().unwrap().remove(&task_id);
    STARTING_TASKS.lock().unwrap().remove(&task_id);
    EVENT_CTX.task_usage.lock().unwrap().remove(&task_id);

    if status == 0 {
        scan_git_info(working_dir, task_id, db);
        tasks::add_log(db, task_id, "Claude finished successfully.", "success", None);
        tasks::update_status(db, task_id, "testing");
        tasks::set_completed(db, task_id);
        if let Some(updated) = tasks::get_by_id(db, task_id) {
            app.emit("task:updated", &updated).ok();
        }
        activity::add(db, project_id, Some(task_id), "task_completed", &format!("Task completed: {}", task_title), None);
    } else {
        tasks::add_log(db, task_id, &format!("Claude exited with code {}.", status), "error", None);
        activity::add(db, project_id, Some(task_id), "task_failed", &format!("Task failed (exit {}): {}", status, task_title), None);
        crate::services::queue::handle_task_failure(db, app, project_id, task_id);
    }

    // Cleanup attachments
    if attach_dir.exists() {
        std::fs::remove_dir_all(attach_dir).ok();
    }

    app.emit("claude:finished", &serde_json::json!({"taskId": task_id, "exitCode": status})).ok();
}

pub fn start(
    task: &tasks::Task,
    app: AppHandle,
    working_dir: &str,
    project: &projects::Project,
    mcp_server_port: u16,
) {
    let task_id = task.id;
    let db = db::get_db();

    if is_running(task_id) || STARTING_TASKS.lock().unwrap().contains(&task_id) {
        return;
    }
    STARTING_TASKS.lock().unwrap().insert(task_id);

    // Copy attachments to working dir
    let (task_attachments, attach_dir) = copy_task_attachments(task_id, working_dir, &db);

    let revisions = tasks::get_revisions(&db, task_id);
    let enabled_snippets = snippets::get_enabled_by_project(&db, task.project_id);
    let role = task.role_id.and_then(|rid| roles::get_by_id(&db, rid));

    let prompt = build_prompt(task, &revisions, &enabled_snippets, &task_attachments, role.as_ref(), task.project_id);
    let model = task.model.as_deref().unwrap_or("sonnet");
    let effort = task.thinking_effort.as_deref().unwrap_or("medium");
    let permission_mode = project.permission_mode.as_deref().unwrap_or("auto-accept");
    let allowed_tools = project.allowed_tools.as_deref().unwrap_or("");

    // Snapshot baseline usage
    if let Some(current) = tasks::get_by_id(&db, task_id) {
        EVENT_CTX.task_usage.lock().unwrap().insert(task_id, UsageTracker {
            baseline: UsageBaseline {
                input: current.input_tokens.unwrap_or(0),
                output: current.output_tokens.unwrap_or(0),
                cache_read: current.cache_read_tokens.unwrap_or(0),
                cache_creation: current.cache_creation_tokens.unwrap_or(0),
                cost: current.total_cost.unwrap_or(0.0),
            },
            session: UsageSession::default(),
        });
    }

    // Auto-create branch
    let mut task_clone = task.clone();
    if let Some(branch) = ensure_task_branch(task, working_dir, project, &db, &app) {
        task_clone.branch_name = Some(branch);
    }

    tasks::add_log(&db, task_id, &format!("Starting Claude for task: {}", task.title), "system", None);
    tasks::add_log(&db, task_id, &format!("Model: {} | Effort: {} | Permissions: {}", model, effort, permission_mode), "info", None);
    activity::add(&db, task.project_id, Some(task_id), "claude_started", &format!("Claude started: {}", task.title), None);

    // Build CLI arguments
    let args = build_claude_args(&prompt, model, effort, permission_mode, allowed_tools, mcp_server_port);

    let working_dir = working_dir.to_string();
    let project_id = task.project_id;
    let task_title = task.title.clone();

    std::thread::spawn(move || {
        let mut cmd = Command::new("claude");
        cmd.args(&args)
            .current_dir(&working_dir)
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .stdin(Stdio::null());
        #[cfg(target_os = "windows")]
        cmd.creation_flags(CREATE_NO_WINDOW);

        let child = match cmd.spawn() {
            Ok(c) => c,
            Err(e) => {
                let db = db::get_db();
                tasks::add_log(&db, task_id, &format!("Failed to start Claude: {}", e), "error", None);
                STARTING_TASKS.lock().unwrap().remove(&task_id);
                app.emit("claude:finished", &serde_json::json!({"taskId": task_id, "exitCode": -1})).ok();
                return;
            }
        };

        let db = db::get_db();
        handle_process_lifecycle(task_id, child, &db, &app, &working_dir, project_id, &task_title, &attach_dir);
    });
}
